unified_planning/engines/compilers/trajectory_constraints_remover.py

```diff
@@ -0,0 +1,422 @@
+                        env, c.args[0], a
+                    )
+                elif c.is_at_most_once():
+                    precondition, to_add = self._manage_amo_compilation(
+                        env, c.args[0], self._monitoring_atom_dict[c], a, E
+                    )
+                elif c.is_sometime_before():
+                    precondition, to_add = self._manage_sb_compilation(
+                        env, c.args[0], c.args[1], self._monitoring_atom_dict[c], a, E
+                    )
+                elif c.is_sometime():
+                    self._manage_sometime_compilation(
+                        env, c.args[0], self._monitoring_atom_dict[c], a, E
+                    )
+                elif c.is_sometime_after():
+                    self._manage_sa_compilation(
+                        env, c.args[0], c.args[1], self._monitoring_atom_dict[c], a, E
+                    )
+                else:
+                    raise Exception(
+                        f"ERROR This compiler cannot handle this constraint = {c}"
+                    )
+                if c.is_always() or c.is_at_most_once() or c.is_sometime_before():
+                    if to_add and not precondition.is_true():
+                        a.preconditions.append(precondition)
+            for eff in E:
+                a.effects.append(eff)
+            if env.expression_manager.FALSE() not in a.preconditions:
+                A_prime.append(a)
+            trace_back_map[a] = map_value
+        # create new problem to return
+        # adding new fluents, goal, initial values and actions
+        G_new = (env.expression_manager.And(self._problem.goals, G_prime)).simplify()
+        self._problem.clear_goals()
+        self._problem.add_goal(G_new)
+        self._problem.clear_trajectory_constraints()
+        for fluent in F_prime:
+            self._problem.add_fluent(fluent)
+        self._problem.clear_actions()
+        for action in A_prime:
+            self._problem.add_action(action)
+        for init_val in I_prime:
+            self._problem.set_initial_value(
+                up.model.Fluent(f"{init_val}", env.type_manager.BoolType()), True
+            )
+        return CompilerResult(
+            self._problem, partial(lift_action_instance, map=trace_back_map), self.name
+        )
+
+    def _build_constraint_list(self, expression_quantifier_remover, env, constraints):
+        C_temp = (env.expression_manager.And(constraints)).simplify()
+        C_list = C_temp.args if C_temp.is_and() else [C_temp]
+        C_to_return = (
+            env.expression_manager.And(
+                self._remove_quantifier(expression_quantifier_remover, C_list)
+            )
+        ).simplify()
+        return C_to_return.args if C_to_return.is_and() else [C_to_return]
+
+    def _remove_quantifier(self, expression_quantifier_remover, C):
+        new_C = []
+        for c in C:
+            assert c.node_type is not OperatorKind.EXISTS
```

Comment:
Why not?

And also, what about an expression like `And(x, Exists(y, And(x, y))`?

Reply:
trajectory constraints support Exists only outside.

Reply:
@hstairs My understanding was that `Exists` is not supported "outside", but "inside" it should be allowed (as it amounts to a simple disjunction).

Reply:
@mikand yes "inside" is allowed. In my test "_remove_quantifier" remove internal exist with no error.
For example `(always(and (at a)(exists(?x)(at ?x))))` is compiled in a correct way in `Always((at(a) and (at(a) or at(b) or at(c))))`